Lpc 11u6x pinctrl

danieldegrasse opened the pull request:
Enable pin control on LPC11u6x SOCs. The following boards have been converted:
- faze
- lpcxpresso11u68

~This PR is dependent on #44842, as the LPC11u6x port configuration registers are not contiguous, so calculating offsets using the port and pin is not possible.~

danieldegrasse committed:
dts: lpc11u6x: update lpc dtsi file to include pin control nodes

update lpc11u6x dtsi file to include required DTS nodes to support LPC
pin control

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: lpc11u6x: add pin control definitions to LPC11u6x soc file

add pin control definitions to LPC11u6x soc file, to handle the lack of
a HAL for this SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: gpio_lpc11u6x: use pio nodes to configure pin mux for gpio

switch gpio driver to use pio nodes to configure pin control settings,
and stop using pinmux driver within gpio driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pinctrl: update lpc pinctrl driver for lpc11u6x

Update pin control driver for lpc11u6x. This SOC does not have a HAL,
so fsl_clock is not available. It also lacks a slew-rate field in the
IOCON register, so this property must be optional.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: faze: add pin control nodes to faze board

add pin control definitions and nodes to faze board, which uses an
LPC11u67 SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: lpc11u6x: enable pin control

Enable pin control for lpc11u6x soc by selecting CONFIG_PINCTRL=y.
At this time no drivers are ported.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: clock_control: convert lpc11u6x syscon driver to pinctrl

convert lpc11u6x syscon clock driver to pin control, and remove all
pinmux usage from driver and syscon dts node.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: i2c: i2c_lpc11u6x: enable pin control

Enable pin control for lpc11u6x i2c driver, and remove pinmux usage from
board level DTS files.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: serial: enable pin control for lpc11u6x serial driver

Enable pin control api for lpc11u6x serial driver, and remove pinmux api
usage.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: lpcxpresso11u68: add pin control nodes to lpcxpresso11u68

Add pin control node definitions to lpcxpresso11u68 board.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
include: dt-bindings: add lpc pin control definitions

add pin control definitions for LPC11u6x to LPU11u6x pinctrl binding.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@mbittan I am unsure if you are still active on Github, but could you test this PR against an LPCxpresso11u68 if possible?

The project has a mandate to move from pinmux APIs to Linux style pin control APIs to handle pin selection by the next release, hence these changes. I do not currently have hardware to test on, so I can only verify that this patch builds for the faze and lpcxpresso11u68 targets. If the serial port input/output works correctly, as well as an application using I2C, then this PR likely works fine on hardware.

mbittan commented:
> @mbittan I am unsure if you are still active on Github, but could you test this PR against an LPCxpresso11u68 if possible?
> 
> The project has a mandate to move from pinmux APIs to Linux style pin control APIs to handle pin selection by the next release, hence these changes. I do not currently have hardware to test on, so I can only verify that this patch builds for the faze and lpcxpresso11u68 targets. If the serial port input/output works correctly, as well as an application using I2C, then this PR likely works fine on hardware.

I will test this as soon as possible

danieldegrasse committed:
soc: lpc11u6x: add pin control header

add pin control header to enable pin control support for lpc11u6x

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbittan commented on drivers/serial/uart_lpc11u6x.c:
```diff
@@ -874,25 +855,14 @@ static const struct uart_driver_api uartx_api = {
 
 
 #define LPC11U6X_UARTX_INIT(idx)                                              \
+PINCTRL_DT_INST_DEFINE(idx);                                                  \
```
We can't use the instance number for this driver because all the UART controllers of the LPC11U6X are not the same.
This should work:
 ```suggestion
PINCTRL_DT_DEFINE(DT_NODELABEL(uart##idx));                                   \
```

mbittan commented on drivers/serial/uart_lpc11u6x.c:
```diff
@@ -874,25 +855,14 @@ static const struct uart_driver_api uartx_api = {
-	.tx_pin = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs, txd, pin), \
-	.tx_func = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs,	      \
-				  txd, function),			      \
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),                        \
```
The instance number can't be used here either, this should work:
```suggestion
        .pincfg = PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(uart##idx)),         \
```

mbittan commented on drivers/serial/uart_lpc11u6x.h:
```diff
@@ -131,10 +133,7 @@ struct lpc11u6x_uart0_config {
 	const char *tx_pinmux_drv_name;
```
This can be removed too

mbittan reviewed (changes requested):
I tested this PR on a lpcxpresso11u68 board: everything is ok except UART which does not work. The changes I suggest fix the issue on both boards.

danieldegrasse commented:
@mbittan I've implemented your changes, but when testing using the LPCXpresso11u68 EVK we are not seeing output on the UART4 pins. This issue exists in this PR, and on the `main` branch. Is there a step we are missing in testing the Zephyr enablement for this board? Also, what was the reason for selecting UART4 as the primary Zephyr console, instead of UART0, which is connected to the onboard debugger?

mbittan commented:
> @mbittan I've implemented your changes, but when testing using the LPCXpresso11u68 EVK we are not seeing output on the UART4 pins. This issue exists in this PR, and on the `main` branch. Is there a step we are missing in testing the Zephyr enablement for this board? 

UART4 should work without any specific configuration. For example, the hello_world sample works correctly on my board (which is an LPCxpressoV2 RevC) with your updated patches and on the main branch. What revision of the board are you using ? How are you testing the UART4 on your board ?
 
> Also, what was the reason for selecting UART4 as the primary Zephyr console, instead of UART0, which is connected to the onboard debugger?

We chose UART4 because it was accessible on the arduino connector, but I agree that it makes more sense to use UART0.

Regarding the LPC11U6*-pinctrl.h files, I think it would be better it they were located in the zephyr tree. That would allow the lpc11u6x port to stay independent of the NXP HAL.

simonguinot commented:
> Regarding the LPC11U6*-pinctrl.h files, I think it would be better it they were located in the zephyr tree. That would allow the lpc11u6x port to stay independent of the NXP HAL.

Hi @danieldegrasse.

I agree with Maxime that there may be rewards (don't ask me which ones, I'd be embarrassed) to keep the lpc11u6x support standalone (i.e. not dependent on the NXP HAL). It was part of the effort when we wrote this port.

I think that the pin muxing is the same for all LPC11U6x SoCs. The only difference is the availability of pins which depends on the package size (48, 64 or 100 pins). So, IMHO, we can have a single pinctrl header file in Zephyr to hold all the definitions.

Note that the LPC11U67JBD48-pinctrl.h and LPC11U68JBD100-pinctrl.h files provided by the NXP HAL are identical. And so the LPC11U67JBD48-pinctrl.h file does include definitions for pins that are not available on the 48-pin package...

danieldegrasse committed:
boards: lpcxpresso11u68: change default console to uart0

UART0 is routed to onboard debugger on LPC11u68 EVK. Change the default
console from UART4 to UART0 to enable output on the onboard debugger
console.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

simonguinot commented on boards/arm/faze/faze-pinctrl.dtsi:
```diff
@@ -0,0 +1,66 @@
+				<I2C0_SDA_PIO0_5>;
+		};
+		nxp,i2c-mode;
+		nxp,i2c-filter = "slow";
```
I think this properties should be in the group0 node. Else they are not handled by the `Z_PINCTRL_STATE_PIN_INIT` macro you added in `soc/arm/nxp_lpc/lpc11u6x/pinctrl_soc.h`.

This comment also applies to most of the nodes below.

danieldegrasse replied:
Should be fixed in latest push-my bad for missing that.

simonguinot commented on boards/arm/lpcxpresso11u68/lpcxpresso11u68-pinctrl.dtsi:
```diff
@@ -0,0 +1,50 @@
+				<I2C0_SDA_PIO0_5>;
+		};
+		nxp,i2c-mode;
+		nxp,i2c-filter = "slow";
```
I think this properties should be in the group0 node. Else they are not handled by the `Z_PINCTRL_STATE_PIN_INIT` macro you added in `soc/arm/nxp_lpc/lpc11u6x/pinctrl_soc.h`.

The same comment applies to most of the nodes below.

danieldegrasse replied:
Ditto on above-should be fixed now

simonguinot commented on boards/arm/lpcxpresso11u68/lpcxpresso11u68-pinctrl.dtsi:
```diff
@@ -0,0 +1,50 @@
+		nxp,i2c-mode;
+		nxp,i2c-filter = "fast";
+	};
+
```
You have an extra new line here.

danieldegrasse replied:
Fixed, thanks

simonguinot commented on boards/arm/faze/faze-pinctrl.dtsi:
```diff
@@ -0,0 +1,66 @@
+	};
+
+
+
```
You have extra new lines here.

danieldegrasse replied:
Fixed in latest push

simonguinot reviewed (changes requested):
Hi @danieldegrasse,

Thanks for this PR. It is looking good.

I started to review it. Please find my first comments.

Simon

danieldegrasse commented:
DNM until I have had a chance to address requested changes, since @simonguinot's change request won't block merging

mbittan commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -141,23 +139,27 @@ static int gpio_lpc11u6x_pin_configure(const struct device *port,
 
 	if (flags & GPIO_PULL_UP) {
-		func |= IOCON_MODE_PULLUP;
+		func |= IOCON_PIO_MODE(0x1);
```
The correct value for the MODE field is 0x2 to enable pull-up resistor

danieldegrasse replied:
Fixed in latest push

mbittan commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -141,23 +139,27 @@ static int gpio_lpc11u6x_pin_configure(const struct device *port,
+		func |= IOCON_PIO_MODE(0x1);
 	} else if (flags & GPIO_PULL_DOWN) {
-		func |= IOCON_MODE_PULLDOWN;
+		func |= IOCON_PIO_MODE(0x2);
```
The value for the pull-down resistor is 0x1

danieldegrasse replied:
Fixed in latest push

mbittan commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -141,23 +139,27 @@ static int gpio_lpc11u6x_pin_configure(const struct device *port,
+		func |= IOCON_PIO_MODE(0x2);
 	} else {
-		func |= IOCON_MODE_INACT;
+		func |= IOCON_PIO_MODE(0x3);
```
The value to disable pull-up/pull-down resistors is 0x0

danieldegrasse replied:
Fixed in latest push

dleach02 commented:
@mbittan can we get your feedback on the current state of this PR? I would like to get it ready and merged before the feature freeze on 5/13

mbittan reviewed (dismissed):
Looks good to me

gmarull commented on dts/arm/nxp/nxp_lpc11u6x.dtsi:
```diff
@@ -52,6 +52,32 @@
 			status = "okay";
 		};
 
+		iocon: iocon@40044000 {
```
why not name this one `pinctrl`?

danieldegrasse replied:
The `pio0` through `pio2` nodes are used by the GPIO driver to determine base addresses for each IOCON port, so the `compatible` string for the `iocon` node needs to define the properties for each `pio` child node. Then, the child `pinctrl` node has its own properties, and that node has the pincontrol selection defined as children.

I believe if this node was `pinctrl` then the `pio` child nodes could not be present, because they do not contain the same properties as pin control groups (which are the children of the `pinctrl` node) do. Please correct me if that is wrong, because this node being named `pinctrl` would be cleaner.

gmarull replied:
If a child has a compatible set, then it can do whatever it wants, see for example STM32/GD32 pinctrl/gpio layout.

danieldegrasse replied:
Makes sense. The only other hurdle to renaming that pin control node would be that the lpc pin control driver currently relies on the node label `iocon` being present in the devicetree in order to get the base address of the pin controller. I used the node label directly because like the iMX pin control, there are several different register layouts present across the SOC families with the IOCON peripheral, so there is not one shared DTS compatible string that could be used for all pin control nodes.

So I could either:
- rename the `iocon` node for all devices currently using this pin control driver (and rework the dts binding file (currently the RT600 and all LPC SOCs use this driver)
- add a corner case to this driver for the LPC11u6x

personally, I would rather keep the driver and node naming as it currently is designed, but if you feel strongly about this, I can rename the `iocon` node across the LPC line. I would prefer not to add a corner case for the LPC11u6x, however.

gmarull replied:
I think it would make more sense to make `iocon` the pinctrl (following stm32/gd32 layout), but this can be improved later.

gmarull commented on dts/bindings/pinctrl/nxp,lpc-iocon-pinctrl.yaml:
```diff
@@ -125,3 +134,34 @@ child-binding:
+          - 1
+          - 2
+          - 3
+        description: |
+          Enable digital filter. Set number of clock cycles to use as rejection
+          threshold for input pulses. 0 disables the filter. Only valid for
+          lpc11u6x SOC
```
default needs to be justified

danieldegrasse replied:
Fixed in latest push

gmarull commented on dts/bindings/pinctrl/nxp,lpc-iocon-pinctrl.yaml:
```diff
@@ -125,3 +134,34 @@ child-binding:
+          - 4
+          - 5
+          - 6
+        description: |
+          set peripheral clock divider for input filter sampling clock
+          IOCONCLKDIV. Only valid for lpc11u6x SOC
```
ditto

danieldegrasse replied:
Fixed in latest push

gmarull commented on dts/bindings/pinctrl/nxp,lpc-iocon-pinctrl.yaml:
```diff
@@ -125,3 +134,34 @@ child-binding:
+          - 4
+          - 5
+          - 6
+        description: |
+          set peripheral clock divider for input filter sampling clock
+          IOCONCLKDIV. Only valid for lpc11u6x SOC
```
ditto

gmarull commented on dts/bindings/pinctrl/nxp,lpc-iocon-pinctrl.yaml:
```diff
@@ -125,3 +134,34 @@ child-binding:
+        description: |
+          set peripheral clock divider for input filter sampling clock
+          IOCONCLKDIV. Only valid for lpc11u6x SOC
+      nxp,analog-filter:
+        type: boolean
+        description: |
+          Enable fixed 10 ns input glitch analog filter. Only valid for lpc11u6x
+          SOC, on analog pins.
```
ditto

gmarull commented on dts/bindings/pinctrl/nxp,lpc-iocon-pinctrl.yaml:
```diff
@@ -125,3 +134,34 @@ child-binding:
+          - 2
+          - 3
+        description: |
+          Enable digital filter. Set number of clock cycles to use as rejection
+          threshold for input pulses. 0 disables the filter. Only valid for
+          lpc11u6x SOC
```
note: it looks like you want to have multiple compatibles (with a file containing common fields), this way only lpc11u6x SoCs will get these properties exposed.

danieldegrasse replied:
You're right, this is better. I split the LCP11u6x specific fields into a new file.

gmarull commented on boards/arm/faze/faze-pinctrl.dtsi:
```diff
@@ -0,0 +1,62 @@
+#include <dt-bindings/pinctrl/lpc11u6x-pinctrl.h>
+
+&pinctrl {
+	pinmux_uart0: pinmux_uart0 {
```
suggestion: name using target state, e.g. `uart0_default`, `i2c0_fast`, etc.

danieldegrasse replied:
Fixed to use target state naming

gmarull commented on drivers/i2c/i2c_lpc11u6x.c:
```diff
@@ -34,8 +32,9 @@ static int lpc11u6x_i2c_configure(const struct device *dev,
-	const struct device *clk_dev, *pinmux_dev;
-	uint32_t speed, flags = 0;
+	const struct device *clk_dev;
+	uint32_t speed, mux_selection = PINCTRL_STATE_DEFAULT;
```
nit: state identifiers can use `uint8_t`

gmarull commented on dts/bindings/serial/nxp,lpc11u6x-uart.yaml:
```diff
@@ -14,12 +14,12 @@ properties:
+    pinctrl-0:
+      type: phandles
       required: true
-      description: RX and TX pinmux
+      description: Pinmux selection for RX and TX pins
 
-    pinmux-names:
+    pinctrl-names:
       type: string-array
       required: true
-      description: UART pins names
+      description: Pin control names
```
no need to re-define type/description for `pinctrl-` fields, they're defined in `pinctrl-device.yaml` (check other cases as well).

danieldegrasse replied:
Dropped these re-definitions where they were present

simonguinot reviewed (dismissed):
Hi @danieldegrasse,
Thanks for this PR. It is looking good.

gmarull reviewed (changes requested):
please update includes, otherwise lgtm

danieldegrasse commented:
@gmarull headers should be updated
